src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java

```diff
@@ -0,0 +1,255 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.DeletionTime;
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.db.rows.Unfiltered;
+import org.apache.cassandra.db.tries.InMemoryTrie;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeyFactory;
+import org.apache.cassandra.io.sstable.format.SSTableFlushObserver;
+
+/**
+ * Writes all on-disk index structures attached to a given SSTable.
+ */
+public class StorageAttachedIndexWriter implements SSTableFlushObserver
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final IndexDescriptor indexDescriptor;
+    private final PrimaryKeyFactory primaryKeyFactory;
+    private final Collection<StorageAttachedIndex> indexes;
+    private final Collection<PerIndexWriter> perIndexWriters;
+    private final PerSSTableWriter perSSTableWriter;
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+    private final RowMapping rowMapping;
+
+    private DecoratedKey currentKey;
+    private boolean tokenOffsetWriterCompleted = false;
+    private boolean aborted = false;
+
+    private long sstableRowId = 0;
+
+    public StorageAttachedIndexWriter(IndexDescriptor indexDescriptor,
+                                      Collection<StorageAttachedIndex> indexes,
+                                      LifecycleNewTracker lifecycleNewTracker) throws IOException
+    {
+        this(indexDescriptor, indexes, lifecycleNewTracker, false);
+    }
+
+    public StorageAttachedIndexWriter(IndexDescriptor indexDescriptor,
+                                      Collection<StorageAttachedIndex> indexes,
+                                      LifecycleNewTracker lifecycleNewTracker,
+                                      boolean perIndexComponentsOnly) throws IOException
+    {
+        this.indexDescriptor = indexDescriptor;
+        this.primaryKeyFactory = indexDescriptor.primaryKeyFactory;
```

Comment:
nit: Not that it's super costly, but given we have `indexDescriptor`, we could probably do without the separate `primaryKeyFactory` field.